DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Services/TransactionManagementService.cs

```diff
@@ -0,0 +1,249 @@
+/// This includes starting, pausing, committing, and rolling back transactions
+/// </summary>
+public sealed class TransactionManagementService : ITransactionManagementService
+{
+  private readonly Lazy<RevitConversionContextStack> _lazyContextStack;
+  private readonly ErrorPreprocessingService _errorPreprocessingService;
+  private Document Document => _lazyContextStack.Value.Current.Document.Document;
+
+  public TransactionManagementService(
+    ErrorPreprocessingService revitErrorPreprocessingService,
+    Lazy<RevitConversionContextStack> lazyContextStack
+  )
+  {
+    _errorPreprocessingService = revitErrorPreprocessingService;
+    _lazyContextStack = lazyContextStack;
+  }
+
+  private TransactionGroup? _transactionGroup;
+  private Transaction? _transaction;
+  private SubTransaction? _subTransaction;
+
+  public void StartTransactionManagement(string transactionName)
+  {
+    if (_transactionGroup == null)
+    {
+      _transactionGroup = new TransactionGroup(Document, transactionName);
+      _transactionGroup.Start();
+    }
+    StartTransaction();
+  }
+
+  public void FinishTransactionManagement()
+  {
+    try
+    {
+      CommitTransaction();
+    }
+    finally
+    {
+      if (_transactionGroup?.GetStatus() == TransactionStatus.Started)
+      {
+        _transactionGroup.Assimilate();
+      }
+      _transactionGroup?.Dispose();
+    }
+  }
+
+  public void RollbackTransactionManagement()
+  {
+    RollbackTransaction();
+    if (
+      _transactionGroup != null
+      && _transactionGroup.IsValidObject
+      && _transactionGroup.GetStatus() == TransactionStatus.Started
+    )
+    {
+      _transactionGroup.Assimilate();
+    }
+  }
+
+  public void StartTransaction()
+  {
+    if (_transaction == null || !_transaction.IsValidObject || _transaction.GetStatus() != TransactionStatus.Started)
+    {
+      _transaction = new Transaction(Document, "Speckle Transaction");
+      var failOpts = _transaction.GetFailureHandlingOptions();
+      failOpts.SetFailuresPreprocessor(_errorPreprocessingService);
+      failOpts.SetClearAfterRollback(true);
+      _transaction.SetFailureHandlingOptions(failOpts);
+      _transaction.Start();
+    }
+  }
+
+  public TransactionStatus CommitTransaction()
+  {
+    if (
+      _subTransaction != null
+      && _subTransaction.IsValidObject
+      && _subTransaction.GetStatus() == TransactionStatus.Started
+    )
+    {
+      HandleFailedCommit(_subTransaction.Commit());
+      _subTransaction.Dispose();
+    }
+    if (_transaction != null && _transaction.IsValidObject && _transaction.GetStatus() == TransactionStatus.Started)
+    {
+      var status = _transaction.Commit();
+      HandleFailedCommit(status);
+      _transaction.Dispose();
+      return status;
+    }
+    return TransactionStatus.Uninitialized;
+  }
+
+  private void HandleFailedCommit(TransactionStatus status)
```

Comment:
This feels like it maybe doing too much and is using SpeckleLog directly.

Reply:
you are right. Removed the error handling and added POC comments